lib/src/core/verifier/policy/opa/opa.go

```diff
@@ -0,0 +1,78 @@
+package main
+
+import "C"
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/open-policy-agent/opa/storage/inmem"
+)
+
+//export evaluateGo
+func evaluateGo(policy string, data string, input string) *C.char {
+	// Deserialize the message in json format
+	input_map := make(map[string]interface{})
+	err := json.Unmarshal([]byte(input), &input_map)
+	if err != nil {
+		return C.CString("Error:: " + err.Error())
+	}
+
+	data_map := make(map[string]interface{})
+	err2 := json.Unmarshal([]byte(data), &data_map)
+	if err2 != nil {
+		// Handle error.
+		return C.CString("Error:: " + err.Error())
+	}
+	// Manually create the storage layer. inmem.NewFromObject returns an
+	// in-memory store containing the supplied data.
+	store := inmem.NewFromObject(data_map)
+
+	// Construct a Rego object that can be prepared or evaluated.
+	r := rego.New(
+		rego.Query("input;data.policy"),
+		rego.Module("demo.rego", policy),
+		rego.Store(store),
+	)
+
+	// Create a prepared query that can be evaluated.
+	ctx := context.Background()
+	query, err := r.PrepareForEval(ctx)
+	if err != nil {
+		return C.CString("Error:: " + err.Error())
+	}
+
+	// Make opa query
+	rs, err := query.Eval(ctx, rego.EvalInput(input_map))
+	if err != nil {
+		return C.CString("Error:: " + err.Error())
+	}
+
+	// Transform the processed decision into the format rust hopes for
+	inputOPA := rs[0].Expressions[0].Value.(map[string]interface{})
+	dataOPA := rs[0].Expressions[1].Value.(map[string]interface{})
+	parseInfo := make(map[string]interface{})
+
+	for k, v := range inputOPA {
+		value := [2]interface{}{v, data_map[k]}
+		parseInfo[k] = value
+	}
+
+	decisionMap := make(map[string]interface{})
+	decisionMap["parseInfo"] = parseInfo
+	decisionMap["allow"] = dataOPA["allow"]
```

Comment:
This can be a go struct (with `json` annotations if needed) instead of a free-form map of strings.

Reply:
Yes, use a struct will make the codes clear, but this glue code is straightforward, so I think we needn't add too much optimization here.

Reply:
Yes, you comment reminds me that a bug exist here. This code needs to pass all the whole OPA evaluation results to the RUST, and the evaluation is dynamically according to the `Policy` file. So it shouldn't use a go struct here but I will update the codes as the following:
```
	// OPA's evaluation results.
	for k, v := range dataOPA {
	    decisionMap[k] = v
	}
	// The input and reference data pair.
	decisionMap["parseInfo"] = parseInfo
```
And I will also enhance the `core\verifier\policy\opa\mod.rs` cargo test cases to cover this scenario.